packages/blitz-rpc/src/data-client/rpc.ts

```diff
@@ -101,56 +97,46 @@ export function __internal_buildRpcClient({
+      __BLITZ_onRpcError,
+      __BLITZ_onSessionCreated,
+    } = globalThis
 
     const promise = window
-      .fetch(routePathURL, {
-        method: httpMethod,
-        headers,
-        credentials: "include",
-        redirect: "follow",
-        body:
-          httpMethod === "POST"
-            ? JSON.stringify({
-                params: serialized.json,
-                meta: {
-                  params: serialized.meta,
-                },
-              })
-            : undefined,
-        signal,
-      })
+      .fetch(
+        routePathURL,
+        __BLITZ_beforeHttpRequest({
+          method: httpMethod,
+          headers: {
+            "Content-Type": "application/json",
+          },
+          credentials: "include",
+          redirect: "follow",
+          body:
+            httpMethod === "POST"
+              ? JSON.stringify({
+                  params: serialized.json,
+                  meta: {
+                    params: serialized.meta,
+                  },
+                })
+              : undefined,
+          signal,
+        }),
+      )
       .then(async (response) => {
         debug("Received request for", routePath)
-        if (response.headers) {
-          backupAntiCSRFTokenToLocalStorage()
-          if (response.headers.get(HEADER_PUBLIC_DATA_TOKEN)) {
-            getPublicDataStore().updateState()
-            debug("Public data updated")
-          }
-          if (response.headers.get(HEADER_SESSION_CREATED)) {
-            // This also runs on logout, because on logout a new anon session is created
-            debug("Session created")
-            setTimeout(async () => {
-              // Do these in the next tick to prevent various bugs like https://github.com/blitz-js/blitz/issues/2207
-              debug("Invalidating react-query cache...")
-              await getQueryClient().cancelQueries()
-              await getQueryClient().resetQueries()
-              getQueryClient().getMutationCache().clear()
-              // We have a 100ms delay here to prevent unnecessary stale queries from running
-              // This prevents the case where you logout on a page with
-              // Page.authenticate = {redirectTo: '/login'}
-              // Without this delay, queries that require authentication on the original page
-              // will still run (but fail because you are now logged out)
-              // Ref: https://github.com/blitz-js/blitz/issues/1935
-            }, 100)
-          }
-          if (response.headers.get(HEADER_CSRF_ERROR)) {
-            const err = new CSRFTokenMismatchError()
-            err.stack = null!
-            throw err
-          }
-        }
-
+        document.addEventListener(
+          "blitz:session-created",
+          async () => {
+            await Promise.all(__BLITZ_onSessionCreated(resetQueryClient))
+            debug("blitz-rpc:session-created")
+          },
+          {once: true},
+        )
```

Comment:
this `addEventListener` call must be inside setupBlitzClient instead of here

Reply:
I have moved the `addEventListener` to the `useEffect` in `buildWithBlitz` function in the same file. Cannot add in setupBlitzClient because as far as I understand it runs in the server and `document` does not exist there.
Also, I removed the parameter `resetQueryClient` instead of keeping the definition in `blitz-rpc` and running it there. 
Is this fine? 
Or should I move the `resetQueryClient` function to `blitz-next` ?

Reply:
resetQueryClient is now good.

addEventListener needs moved to setupBlitzClient, see my comment below